scheduler-3/src/main/scala/uk/sky/scheduler/Main.scala

```diff
@@ -1,30 +1,39 @@
 package uk.sky.scheduler
 
-import cats.data.Reader
+import cats.Parallel
 import cats.effect.*
+import cats.effect.kernel.Resource.ExitCase
+import cats.syntax.all.*
 import fs2.Stream
 import org.typelevel
 import org.typelevel.log4cats.LoggerFactory
 import org.typelevel.log4cats.slf4j.Slf4jFactory
+import org.typelevel.otel4s.metrics.Meter
+import org.typelevel.otel4s.oteljava.OtelJava
 import pureconfig.ConfigSource
 import pureconfig.module.catseffect.syntax.*
 import uk.sky.scheduler.config.Config
 
 object Main extends IOApp.Simple {
 
-  def stream: Reader[Config, Stream[IO, Unit]] =
+  def stream[F[_] : Async : Parallel : LoggerFactory : Meter](config: Config): Stream[F, Unit] =
     for {
-      scheduler <- Scheduler.live[IO]
-      stream    <- Reader[Config, Stream[IO, Scheduler[IO, Unit]]](_ => Stream.resource(scheduler))
-    } yield stream.flatMap(_.stream)
+      scheduler <- Stream.resource(Scheduler.live[F](config))
+      message   <- scheduler.stream
+    } yield message
 
   override def run: IO[Unit] = for {
     given LoggerFactory[IO] <- IO(Slf4jFactory.create[IO])
     logger                  <- LoggerFactory[IO].create
-    config                  <- ConfigSource.default.loadF[IO, Config]()
-    _                       <- logger.info(s"Running ${Config.metadata.appName} with version ${Config.metadata.version}")
-    _                       <- logger.info(s"Loaded config: $config")
-    _                       <- stream(config).compile.drain
+    otel4s                  <- OtelJava.global[IO]
+    given Meter[IO]         <- otel4s.meterProvider.get(Config.metadata.appName)
+    config                  <- ConfigSource.default.at(Config.metadata.appName).loadF[IO, Config]()
+    _                       <- logger.info(show"Running ${Config.metadata.appName} with version ${Config.metadata.version}")
+    _                       <- stream[IO](config).onFinalizeCase {
+                                 case ExitCase.Succeeded  => logger.info("Stream Succeeded")
+                                 case ExitCase.Errored(e) => logger.error(e)(s"Stream error - ${e.getMessage}")
+                                 case ExitCase.Canceled   => logger.info("Stream canceled")
+                               }.compile.drain
```

Comment:
```suggestion
   resource                 = Scheduler.live[IO].apply(config).map(_.stream)
    _                       <- resource.use { stream =>
                                 stream
                                   .onFinalizeCase[IO] {
                                     case ExitCase.Succeeded  => logger.info("Stream Succeeded")
                                     case ExitCase.Errored(e) => logger.error(e)(s"Stream error - ${e.getMessage}")
                                     case ExitCase.Canceled   => logger.info("Stream canceled")
                                   }
                                   .compile
                                   .drain
                               }
```

This way you don't need the stream function, since all the implicits are created within the `run` function.

Reply:
addressed in https://github.com/sky-uk/kafka-message-scheduler/pull/373/commits/0c7d047bd667eb5b8a169b7debf4b80c65386264